### PR TITLE
[NEW] Add option to draw something and send

### LINKED ISSUE #825: [NEW] Add option to draw something and send

### PR DESCRIPTION
**Pull request by artrmz**
@RocketChat/ios

Closes #825

- [x] Create drawing controller
- [x] Brush width settings
- [x] Brush opacity settings
- [x] Brush color picker
- [x] Ignore unused working area (blank/white background) while creating UIImage
- [x] Show dialog or disable done button if there is nothing drawn
- [x] Tests

draw

### CODE REVIEW

**Comment by soundstorm** [comment]
- [ ] Convert drawn to ASCII-Art :joy:

**Comment by artrmz** [comment]
@soundstorm Haha, that sounds great. Could do that in another PR.


https://ijoshsmith.com/2015/04/29/creating-ascii-art-in-functional-swift/

**Comment by artrmz** [comment]
I have added a screenshot of what I currently have, looking forward for some feedbacks.

**Comment by rafaelks** [comment]
@artrmz This is looking awesome!!! Thank you! Could you help us getting some icons to the screen @arthurgiustina? This way we avoid having the strings in the bottom.

**Comment by rafaelks** [comment]
@artrmz FYI: the PR is looking great and we're going to merge it soon, just waiting for new icons in order to merge it ready to ship 👍

**Comment by artrmz** [comment]
@rafaelks Great :) I might improve color picker in the meantime if I find enough time to do that - if not, I will just do that in another PR I guess.

**Comment by artrmz** [comment]
I have added some free icons and found (will replace once @arthurgiustina is done with some better ones 😄) and improved settings' controllers (added titles to make it clear what the new icons do)

**Comment by arthurgiustina** [comment]
New Icons =)

image

**ZIP FILE:** 
[Icons Draw.zip](https://github.com/RocketChat/Rocket.Chat.iOS/files/1759748/Icons.Draw.zip)

**Comment by artrmz** [comment]
@arthurgiustina Thanks!
@rafaelks Updated icons 😄

File: Rocket.Chat/Resources/de.lproj/Localizable.strings [hunk 1]
```diff
@@ -162,6 +163,15 @@
...
 "chat.message.open_browser" = "Im Browser öffnen";
 
+// Chat Drawing
+"chat.drawing.errorTitle" = "Oops!"; // TODO
+"chat.drawing.messageTitle" = "You have to draw something before sending it."; // TODO
+"chat.drawing.fileName" = "Drawing"; // TODO
+"chat.drawing.settings.opacity" = "Opacity"; // TODO
+"chat.drawing.settings.width" = "Width"; // TODO
+"chat.drawing.settings.color.selected" = "Selected color"; // TODO
+"chat.drawing.settings.color.others" = "Choose color below:"; // TODO
```
Comment (rafaelks) [inline Rocket.Chat/Resources/de.lproj/Localizable.strings]: Some strings aren't being used anymore, like the opacity, width and color, right?

File: Rocket.Chat/Resources/pt-BR.lproj/Localizable.strings [hunk 2]
```diff
@@ -154,6 +154,7 @@
 "chat.upload.shoot_video" = "Gravar um vídeo";
 "chat.upload.choose_from_library" = "Escolher da biblioteca";
 "chat.upload.import_file" = "Importar arquivo...";
+"chat.upload.draw" = "Draw a picture"; // TODO
```
Comment (rafaelks) [inline Rocket.Chat/Resources/pt-BR.lproj/Localizable.strings]: Desenhar algo

File: Rocket.Chat/Resources/en.lproj/Localizable.strings [hunk 3]
```diff
@@ -155,6 +155,7 @@
 "chat.upload.shoot_video" = "Shoot Video";
 "chat.upload.choose_from_library" = "Choose From Library";
 "chat.upload.import_file" = "Import File From...";
+"chat.upload.draw" = "Draw a picture";
```
Comment (rafaelks) [inline Rocket.Chat/Resources/en.lproj/Localizable.strings]: I think I would use "Draw something"

File: Rocket.Chat/Resources/pt-BR.lproj/Localizable.strings [hunk 4]
```diff
@@ -162,6 +163,15 @@
 "chat.message.open_file" = "Toque para abrir";
 "chat.message.open_browser" = "Abrir no Safari";
 
+// Chat Drawing
+"chat.drawing.errorTitle" = "Oops!"; // TODO
+"chat.drawing.messageTitle" = "You have to draw something before sending it."; // TODO
```
Comment (rafaelks) [inline Rocket.Chat/Resources/pt-BR.lproj/Localizable.strings]: Você precisa desenhar algo antes de enviar.

File: Rocket.Chat/Resources/de.lproj/Localizable.strings (hunk 1 above)
Comment (artrmz) [reply Rocket.Chat/Resources/de.lproj/Localizable.strings]: Those are being used here:

zrzut ekranu 2018-02-27 o 18 15 33
zrzut ekranu 2018-02-27 o 18 15 39

Should I remove those? Also, I will change opacity to be percentage (will be better I think)

File: Rocket.Chat/Resources/en.lproj/Localizable.strings (hunk 3 above)
Comment (artrmz) [reply Rocket.Chat/Resources/en.lproj/Localizable.strings]: 👍

File: Rocket.Chat/Resources/de.lproj/Localizable.strings (hunk 1 above)
Comment (rafaelks) [reply Rocket.Chat/Resources/de.lproj/Localizable.strings]: No!!! Keep it, please. I forgot that they were being used here.

Comment (rafaelks) [reply Rocket.Chat/Resources/de.lproj/Localizable.strings]: Yes, percentage sounds better to me also.

**Comment by artrmz** [comment]
@rafaelks Applied your translations. Feel free to translate the last few strings if you want :) 
And yeah, could do that in another PR. Also, I have plans to allow drawing on top of image but probably will have to make it scrollable as well.

File: Rocket.Chat/Resources/de.lproj/Localizable.strings (hunk 1 above)
Comment (soundstorm) [reply Rocket.Chat/Resources/de.lproj/Localizable.strings]: "chat.upload.draw" = "Bild zeichnen";
"chat.drawing.errorTitle" = "Oops!";
"chat.drawing.messageTitle" = "Du musst etwas zeichnen, bevor du es senden kannst.";
"chat.drawing.fileName" = "Zeichnung";
"chat.drawing.settings.opacity" = "Transparenz";
"chat.drawing.settings.width" = "Breite";
"chat.drawing.settings.color.selected" = "Gewählte Farbe";
"chat.drawing.settings.color.others" = "Wähle eine Farbe:";


@rafaelks on web-frontend we changed to the formal "Sie" in german (I'd like to have that switchable), should we change that on mobile too? I liked the informal "Du" more.

**Comment by rafaelks** [comment]
SUPER!!! :-)

**Pull request merged**